lib/cdo/aws/cloudfront.rb

```diff
@@ -0,0 +1,211 @@
+    # `app` is a symbol containing the app name (:pegasus or :dashboard)
+    def self.config(app, reference = nil)
+      config = HTTP_CACHE[app]
+      cloudfront = CONFIG[app]
+      behaviors = config[:behaviors].map do |behavior|
+        if behavior[:path].is_a? Array
+          behavior[:path].map do |path|
+            cache_behavior behavior, path
+          end
+        else
+          cache_behavior behavior, behavior[:path]
+        end
+      end.flatten
+
+      ssl_cert = cloudfront[:ssl_cert]
+      {
+        aliases: {
+          quantity: cloudfront[:aliases].length, # required
+          items: cloudfront[:aliases].empty? ? nil : cloudfront[:aliases],
+        },
+        default_root_object: '',
+        origins: {# required
+          quantity: 1, # required
+          items: [
+            {
+              id: 'cdo', # required
+              domain_name: cloudfront[:origin], # required
+              origin_path: '',
+              custom_origin_config: {
+                http_port: 80, # required
+                https_port: 443, # required
+                origin_protocol_policy: 'match-viewer', # required, accepts http-only, match-viewer
+              },
+            },
+          ],
+        },
+        default_cache_behavior: cache_behavior(config[:default]),
+        cache_behaviors: {
+          quantity: behaviors.length, # required
+          items: behaviors.empty? ? nil : behaviors,
+        },
+        custom_error_responses: {
+          quantity: 0 # required
+        },
+        comment: '', # required
+        logging: {
+          enabled: !!cloudfront[:log], # required
+          include_cookies: false, # required
+          bucket: cloudfront[:log] && "#{cloudfront[:log][:bucket]}.s3.amazonaws.com", # required
+          prefix: cloudfront[:log] && cloudfront[:log][:prefix], # required
+        },
+        price_class: 'PriceClass_All', # accepts PriceClass_100, PriceClass_200, PriceClass_All
+        enabled: true, # required
+        viewer_certificate: ssl_cert ? {
+          # Lookup IAM Certificate ID from server certificate name
+          iam_certificate_id: Aws::IAM::Client.new.
+            get_server_certificate(server_certificate_name: ssl_cert).
+            server_certificate.server_certificate_metadata.server_certificate_id,
+          ssl_support_method: 'sni-only', # accepts sni-only, vip
+          minimum_protocol_version: 'TLSv1', # accepts SSLv3, TLSv1
+          cloud_front_default_certificate: false
+        } : {
+          cloud_front_default_certificate: true,
+          minimum_protocol_version: 'TLSv1' # accepts SSLv3, TLSv1
+        },
+        restrictions: {
+          geo_restriction: {# required
+            restriction_type: 'none', # required, accepts blacklist, whitelist, none
+            quantity: 0 # required
+          },
+        },
+      }.tap do |cf|
+        cf[:caller_reference] = reference || Digest::MD5.hexdigest(Marshal.dump(config)) # required
```

Comment:
Forgot to follow up on this before. Here's the [documentation](http://docs.aws.amazon.com/sdkforruby/api/Aws/CloudFront/Types/DistributionConfig.html#caller_reference-instance_method) for `caller_reference`:

> A unique number that ensures the request can't be replayed. If the CallerReference is new (no matter the content of the DistributionConfig object), a new distribution is created. If the CallerReference is a value you already sent in a previous request to create a distribution, and the content of the DistributionConfig is identical to the original request (ignoring white space), the response includes the same information returned to the original request. If the CallerReference is a value you already sent in a previous request to create a distribution but the content of the DistributionConfig is different from the original request, CloudFront returns a DistributionAlreadyExists error.

As I understand this the caller_reference is only used to prevent replayed requests from causing duplicated side-effects anyway, so there's little-to-no likelihood of something bad happening if the Marshal implementation changes.